Migrate from tbb to onetbb

lkeegan opened the pull request:
- add `FlexbarAtomic`
  - a simple wrapper around `std::atomic` which adds a copy constructor
  - to replace `tbb::atomic` member variables in classes with default copy constructors
- replace removed `tbb::atomic`
  - with `FlexbarAtomic` if copy constructor was assumed
  - with `std::atomic` otherwise
- filters (PairedAlign, PairedInput, PairedOutput)
  - no longer inherit from tbb::filter
  - take and return pointers to actual type instead of void*
  - operator() is now const
- pipeline
  - use parallel_pipeline and make_filter
- use global_control to set max threads
- use `oneapi::tbb` namespace
- add find_package for TBB to CMakeLists.txt

lkeegan committed:
Migrate from tbb to onetbb

- add `FlexbarAtomic`
  - a simple wrapper around `std::atomic` which adds a copy constructor
  - to replace `tbb::atomic` member variables in classes with default copy constructors
- replace removed `tbb::atomic`
  - with `FlexbarAtomic` if copy constructor was assumed
  - with `std::atomic` otherwise
- filters (PairedAlign, PairedInput, PairedOutput)
  - no longer inherit from tbb::filter
  - take and return pointers to actual type instead of void*
  - operator() is now const
- pipeline
  - use parallel_pipeline and make_filter
- use global_control to set max threads
- use `oneapi::tbb` namespace
- add find_package for TBB to CMakeLists.txt

tillea commented:
I've tested this patch with the Debian package and confirm flexbar builds that way.  Unfortunately the test seems to end in some endless loop:

     cd test ; \
     export PATH="/build/flexbar-3.5.0/obj-x86_64-linux-gnu:$PATH" ; \
     echo $PATH ; \
     export HOME=/tmp ; \
     which flexbar ; \
     ./flexbar_test.sh
     /build/flexbar-3.5.0/obj-x86_64-linux-gnu:/usr/sbin:/usr/bin:/sbin:/bin
     /build/flexbar-3.5.0/obj-x86_64-linux-gnu/flexbar
 
     Testing fasta:

Hopefully someone will be able to fire up gdb and find a fix.
Kind regards, Andreas.

lkeegan committed:
Fix parallel_pipeline issues

- pass a lambda with a reference to the filter instead of directly passing the filter to parallel_pipeline
  - passed filters may be copied and/or deleted by tbb
  - flexbar continues to use the filter objects after the pipeline is finished
- call fc.stop() in PairedInput when there is no more input
  - required control flow for first filter in parallel_pipeline that was previously missing

lkeegan commented:
@tillea thanks for testing - I didn't notice the tests folder!
[1c872fa10](https://github.com/seqan/flexbar/pull/41/commits/1c872fa10d474f090633fc95d409aa60607a3f96) should fix the endless loop and another issue that came up with the pipeline changes, and now the tests pass for me

lkeegan committed:
revert accidental removal of pthread linking in src/CMakeLists.txt

lkeegan committed:
use `std::atomic` instead of `FlexbarAtomic` in filters as they are no longer require copy constructors

tillea commented:
This works for me now and I'll upload the Debian package with these patches.  Thanks a lot, Andreas.
